Add colorful tags patch

UtkarshVerma opened the pull request:
This PR adds the colorful tags patch by fitrh: https://github.com/fitrh/dwm/issues/1

I have tested it and it works well. Here's how it looks with alpha, hidevacanttags, and underline patches.
![image](https://user-images.githubusercontent.com/31820255/147406472-85751b68-053c-4258-961b-7e7c06fc5ef0.png)

UtkarshVerma commented on dwm.c:
```diff
@@ -1794,7 +1816,12 @@ drawbarwin(Bar *bar)
 			continue;
+
+		#if BAR_COLORFULTAGS_PATCH
+		drw_setscheme(drw, scheme[SchemeLayout]);
```
I am unsure if this is the right way, @bakkeby

bakkeby commented on patch/bar_tags.c:
```diff
@@ -49,11 +49,23 @@ draw_tags(Bar *bar, BarArg *a)
+			m->tagset[m->seltags] & 1 << i
+			#endif // BAR_UNDERLINETAGS_PATCH
+			? m->colorseltag
+				? tagschemes[i]
```
This is fine as is, but I just wanted to note that if the user configures their build to have, say, 11 tags then dwm will likely crash at some point referring to values that are out of bounds here.

One way to avoid this could be to do something like `tagschemes[i % NUMTAGS]` at least then it would wrap around.

On the same note the entire tagschemes array is redundant in this context as you could just replace the above with

```c
? SchemeTag1 + i
```

(of course you still risk running out of color schemes if increasing the number of tags)

UtkarshVerma replied:
You're right about `tagschemes`, but if we remove that, we'll have to assume that the enums are declared sequentially. Would it be safe to assume that?
Regarding the number of tags, I guess we could mention in the patch description that the user needs to define more colors if they are using more than 9 tags.

bakkeby commented on patches.def.h:
```diff
@@ -155,6 +155,11 @@
+ * https://github.com/fitrh/dwm/issues/1
+ */
+#define BAR_COLORFULTAGS_PATCH 0
+
```
This should be moved out of the bar module section and into the bar options section, probably right beneath BAR_COLOR_EMOJI_PATCH.

UtkarshVerma replied:
I think it makes sense to place it under `BAR_TAGS_PATCH` as it would be easier to navigate.

bakkeby reviewed (commented):
Just a few nitpicks.

UtkarshVerma commented on config.def.h:
```diff
@@ -96,6 +96,14 @@ static const char ptagf[] = "[%s %s]";          /* format of a tag label */
 static const int lcaselbl = 0;                  /* 1 means make tag label lowercase */
 #endif // BAR_TAGLABELS_PATCH
+#if BAR_COLORFULTAGS_PATCH
+static const unsigned int colorseltag= 1;       /* 0 means use SchemeSel for selected tag */
```
Oh, that slipped up. Will correct it.

UtkarshVerma commented on dwm.c:
```diff
@@ -479,6 +492,9 @@ struct Monitor {
 	unsigned int alttag;
 	#endif // BAR_ALTERNATIVE_TAGS_PATCH
+	#if BAR_COLORFULTAGS_PATCH
+	unsigned int colorseltag;
```
exactly, I just wrote it for the sake of the patch. I guess it would be more logical to have it as a global constant.

Apart from that, we could completely remove this variable and just turn it on in the code whenever underline tags are enabled. Since this is dwm-flexipatch we're talking about, I guess it would be reasonable to consider this case.

UtkarshVerma replied:
@bakkeby please let me know if we should keep this variable. I personally don't see any reason for it being there. I just included it for the sake of the patch.

UtkarshVerma commented on config.def.h:
```diff
@@ -290,6 +309,19 @@ static char *colors[][ColCount] = {
 	[SchemeHidSel]       = { hidselfgcolor,    hidselbgcolor,    c000000,              c000000 },
 	[SchemeUrg]          = { urgfgcolor,       urgbgcolor,       urgbordercolor,       urgfloatcolor },
+	#if BAR_COLORFULTAGS_PATCH
+	[SchemeTag]          = { "#666666",        "#1b1d1e",        c000000 },
```
Those colors are just placeholder. I'm still new to the codebase so I left it like that. I guess we can expect users to override these in their own config.

UtkarshVerma commented on patch/bar_tags.c:
```diff
@@ -49,11 +49,23 @@ draw_tags(Bar *bar, BarArg *a)
+			#endif // BAR_UNDERLINETAGS_PATCH
+			? m->colorseltag
+				? tagschemes[i]
+				: SchemeSel
```
Okay, I'll do that.

UtkarshVerma commented on patch/bar_tags.c:
```diff
@@ -49,11 +49,23 @@ draw_tags(Bar *bar, BarArg *a)
+			? m->colorseltag
+				? tagschemes[i]
+				: SchemeSel
+			: SchemeTag
```
Yeah, that makes more sense.

UtkarshVerma commented on patch/bar_tags.c:
```diff
@@ -49,11 +49,23 @@ draw_tags(Bar *bar, BarArg *a)
-			? SchemeUrg
-			: SchemeTagsNorm
+				? SchemeUrg
+				: SchemeTagsNorm
```
Done

UtkarshVerma commented:
I recently added support for colorful tags in the XRDB_PATCH as well. This is what I added to `config.h`:
```c

#if BAR_COLORFULTAGS_PATCH
static char tag1fgcolor[]                = "#f92672";
static char tag2fgcolor[]                = "#a6e22e";
static char tag3fgcolor[]                = "#f4bf75";
static char tag4fgcolor[]                = "#66d9ef";
static char tag5fgcolor[]                = "#ae81ff";
static char tag6fgcolor[]                = "#f8f8f2";
static char tag7fgcolor[]                = "#75715e";
static char tag8fgcolor[]                = "#f4bf75";
static char tag9fgcolor[]                = "#a1efe4";

static char tag1bgcolor[]                = "#272822";
static char tag2bgcolor[]                = "#272822";
static char tag3bgcolor[]                = "#272822";
static char tag4bgcolor[]                = "#272822";
static char tag5bgcolor[]                = "#272822";
static char tag6bgcolor[]                = "#272822";
static char tag7bgcolor[]                = "#272822";
static char tag8bgcolor[]                = "#272822";
static char tag9bgcolor[]                = "#272822";

static char layoutfgcolor[]				 = "#ae81ff";
static char layoutbgcolor[]				 = "#272822";
#endif


// inside colors variable
	#if BAR_COLORFULTAGS_PATCH
	[SchemeTag1]       = { tag1fgcolor,		tag1bgcolor,	c000000 },
	[SchemeTag2]       = { tag2fgcolor,		tag2bgcolor,	c000000 },
    [SchemeTag3]       = { tag3fgcolor,		tag3bgcolor,	c000000 },
    [SchemeTag4]       = { tag4fgcolor,		tag4bgcolor,	c000000 },
    [SchemeTag5]       = { tag5fgcolor,		tag5bgcolor,		c000000 },
	[SchemeTag6]       = { tag6fgcolor,		tag6bgcolor,	c000000 },
	[SchemeTag7]       = { tag7fgcolor,		tag7bgcolor,	c000000 },
	[SchemeTag8]       = { tag8fgcolor,		tag8bgcolor,	c000000 },
	[SchemeTag9]       = { tag9fgcolor,		tag9bgcolor,	c000000 },
	[SchemeLayout]     = { layoutfgcolor,	layoutbgcolor,	c000000 },
	#endif
```

This is what I added to `xrdb.c`:

```c
				#if BAR_COLORFULTAGS_PATCH
				XRDB_LOAD_COLOR("dwm.tags.fg1", tag1fgcolor);
				XRDB_LOAD_COLOR("dwm.tags.bg1", tag1bgcolor);
				XRDB_LOAD_COLOR("dwm.tags.fg2", tag2fgcolor);
				XRDB_LOAD_COLOR("dwm.tags.bg2", tag2bgcolor);
				XRDB_LOAD_COLOR("dwm.tags.fg3", tag3fgcolor);
				XRDB_LOAD_COLOR("dwm.tags.bg3", tag3bgcolor);
				XRDB_LOAD_COLOR("dwm.tags.fg4", tag4fgcolor);
				XRDB_LOAD_COLOR("dwm.tags.bg4", tag4bgcolor);
				XRDB_LOAD_COLOR("dwm.tags.fg5", tag5fgcolor);
				XRDB_LOAD_COLOR("dwm.tags.bg5", tag5bgcolor);
				XRDB_LOAD_COLOR("dwm.tags.fg6", tag6fgcolor);
				XRDB_LOAD_COLOR("dwm.tags.bg6", tag6bgcolor);
				XRDB_LOAD_COLOR("dwm.tags.fg7", tag7fgcolor);
				XRDB_LOAD_COLOR("dwm.tags.bg7", tag7bgcolor);
				XRDB_LOAD_COLOR("dwm.tags.fg8", tag8fgcolor);
				XRDB_LOAD_COLOR("dwm.tags.bg8", tag8bgcolor);
				XRDB_LOAD_COLOR("dwm.tags.fg9", tag9fgcolor);
				XRDB_LOAD_COLOR("dwm.tags.bg9", tag9bgcolor);
				XRDB_LOAD_COLOR("dwm.tags.layoutfg", layoutfgcolor);
				XRDB_LOAD_COLOR("dwm.tags.layoutbg", layoutbgcolor);
				#endif
```

> I would have preferred to commit these to the branch but my tree is kind of a mess right now and I don't have time.

UtkarshVerma commented:
@bakkeby any updates on this?

bakkeby commented:
Updates in terms of merging this in you mean? 

I do not have any immediate plans for taking this for various reasons.
